### ISSUE TITLE
Multiple `http.request`s to the same url cannot be differentiated

### ISSUE THREAD

**Issue by fatboychummy** [SquidDev-CC/CC-Tweaked#629]
When making multiple requests using `http.request` to the same url, the response given in the `http_success` / `http_failure` events cannot be differentiated, causing some problems.

As an example, using my [KristWrap](https://github.com/Fatboychummy-CC/KristWrap) along with its recommended json/sha libraries, I can send two POST requests to krist's 'make a transaction' endpoint. If one of those fails, which failed? I cannot know.

Using the following code which sends two async requests and then listens for responses, we can show this problem in action:
```
local uri = "https://krist.ceriat.net/transactions/"
local kw = require "KristWrap"
local json = require "json"
local pkey = kw.toKristWalletFormat(pkey) -- pkey removed for obvious reasons

local headers = {["content-type"] = "Application/json"}

local fail = json.encode {
  privatekey = pkey,
  to = "kbielbeajd",
  amount = "fdsklf",
  metadata = "Test"
}

local pass = json.encode {
  privatekey = pkey,
  to = "kbielbeajd",
  amount = 1,
  metadata = "Test"
}

local function writeFile(data)
  local i = 0
  local format = "output%d"
  repeat i = i + 1 until not fs.exists(string.format(format, i))

  local h = fs.open(string.format(format, i), 'w')
  h.write(data)
  h.close()
end

local function Read()
  local count = 0
  while true do
    local event, uuu, h = os.pullEvent()
    if event == "http_success" then
      local data = h.readAll()
      local code = h.getResponseCode()

      local combined = string.format("RESPONSE CODE: %d\n\n%s", code, data)
      writeFile(data)
      print("Receive", i)
      count = count + 1
    elseif event == "http_failure" then
      print("Fail response", i)
      writeFile("Failure to send.")
      count = count + 1
    end

    if count >= 2 then break end
  end
end

local function Post()
  local failed = false
  if math.random(0, 1) == 0 then
    http.request(uri, pass, headers)
  else
    http.request(uri, fail, headers)
    failed = true
  end
  print("Send 1")

  if failed then
    http.request(uri, pass, headers)
  else
    http.request(uri, fail, headers)
  end
  print("Send 2")
end

parallel.waitForAll(Read, Post)
```
And responses, `output1`:
```
{"ok":false,"error":"invalid_parameter","parameter":"amount"}
```
`output2`:
```
{"ok":true,"transaction":{"id":1910640,"from":"kpcispvv6p","to":"kbielbeajd","value":1,"time":"2020-12-26T22:46:30.611Z","metadata":"Test","type":"transfer"}}
```

We can see one of them failed, but which one? We can assume the first response coordinates with the first request, but what if timing things happened and the first response was actually to the second request?

I recommend adding a numerical id, returned by `http.request` that we can use in a comparison from `http_success` to determine the specific request id something is responding to, for example:

```
local idreq = http.request("http://whatever.whatever")
local event, url, handle, idresp = os.pullEvent("http_success")
if idreq == idresp then
  -- yay we've got the right one
end
```

**Comment by Lemmmy** [comment]
For what it's worth, the way we usually solve it (and have in other Krist implementations) is by using URL fragments to attach an incrementing/random identifier to the end of the URL, like:

```lua
local url = "https://krist.ceriat.net/addresses#123"
http.request(url)

local event, u = os.pullEvent("http_success")
if u == url then
  -- This is the correct request
end
```

Of course, in a serious implementation you'd probably use a table of URLs mapped to response handlers or something.

**Comment by SquidDev** [comment]
> I recommend adding a numerical id, returned by http.request that we can use in a comparison from http_success to determine the specific request id something is responding to, for example:

Yeah, I've been discussing this for years, but really not a fan of it. It probably is the only feasible solution, but chucking more things in the event feels a little ugly too.

**Comment by hugeblank** [comment]
Wouldn’t replacing the URL and adding a numerical ID keep the same number of event arguments?

**Comment by SquidDev** [comment]
> Wouldn’t replacing the URL and adding a numerical ID keep the same number of event arguments? 

Yes, but it would break every program which uses `http.request` directly. :/

**Comment by hugeblank** [comment]
Unless there's a use case for the URL that's returned by the event outside of comparing it with the initial requested URL, I don't think that *every* program that uses it directly would break, would it?

**Comment by fatboychummy** [comment]
> I don't think that _every_ program that uses it directly would break, would it?

Unfortunately a great many would, mine included. I usually "build" my urls and then compare the result to what was returned (allows for quickly changing what url we use by only changing one thing), so like:

```lua
local urlFormat = "https://%s/%s"
local website = "whatever.com"
local endpoint = "transactions/new"
local builtUrl = string.format(urlFormat, website, endpoint)

http.request(builtUrl, --[[...]])

local event, url, body = os.pullevent("http_success")
if url == builtUrl then
  --[[...]]
```

**Comment by CelDaemon** [comment]
It doesn't seem there has been any progress here, but I'm still hoping to see this fixed.

I don't see why it is not possible to simply return a numeric id from `http.request` and providing the same id in `http_success` and `http_failure` (like the timer already does). Although it'd have to be passed as the 4th return value to preserve the functionality of code like in the comment above this one.

It would also not break most existing programs, as the URL can be relied upon the same as before. It'd just allow for more reliable identification of requests, without passing odd hashes.

The current boolean return value would be the biggest issue here, but it is no longer documented and I have not encountered any programs still using it.

**Comment by CelDaemon** [comment]
Also, I feel the same should probably be done for `http.websocketAsync`, as it currently also doesn't have any way to identify the `websocket_success` and `websocket_failure` events, except from checking the URL.

**Comment by CelDaemon** [comment]
I was also thinking that the numeric id of the request could replace the current boolean return value, returning a number on success, and possibly `nil` on failure.

That way, backwards compatibility should be less of a concern, as any programs that `assert` the return value or directly feed it into an it statement will work as normal.

The only issue I can think of, is if a program specifically checks for equality to `false`, but I don't think this is normally done, and the risk should be small enough to safely implement something as important as properly identifiable requests.

**Comment by Wojbie** [comment]
> The only issue I can think of, is if a program specifically checks for equality to `false`, but I don't think this is normally done, and the risk should be small enough to safely implement something as important as properly identifiable requests.

This comparing to true/false is something commonly seen in beginners code. So while this is a good solution it would mean those programs would break.

**Comment by CelDaemon** [comment]
> > The only issue I can think of, is if a program specifically checks for equality to `false`, but I don't think this is normally done, and the risk should be small enough to safely implement something as important as properly identifiable requests.
> 
> This comparing to true/false is something commonly seen in beginners code. So while this is a good solution it would mean those programs would break.

Hm that's understandable, although I haven't seen any programs do that in the wild.
It has also been 3 years since the deprecation of the return values, so I'm hoping those programs, wherever they might be, will likely no longer be in use or fixed to not strictly check for `false`.

Also, if the biggest issue is beginner programs, I haven't seen many of those that check for errors at all.
I think it's more likely that the return value is completely ignored in beginner programs, than it being strictly compared to `false`.

**Comment by SquidDev** [comment]
> I don't see why it is not possible to simply return a numeric id

Oh, it's definitely possible, but as mentioned in https://github.com/cc-tweaked/CC-Tweaked/issues/629#issuecomment-751445744 it just feels incredibly ugly! Given the relatively easy workaround of using the URL fragment to uniquify the URLs, it's been easier to not worry too much about this issue.

We probably should do a better job of mentioning this in the docs. And possibly make `http.get`/`http.post` uniquify automatically behind the scenes.

**Comment by CelDaemon** [comment]
> Oh, it's definitely possible, but as mentioned in [#629 (comment)](https://github.com/cc-tweaked/CC-Tweaked/issues/629#issuecomment-751445744) it just feels incredibly ugly! Given the relatively easy workaround of using the URL fragment to uniquify the URLs, it's been easier to not worry too much about this issue.
> 
> We probably should do a better job of mentioning this in the docs. And possibly make `http.get`/`http.post` uniquify automatically behind the scenes.

It's definitely good to mention it in the docs, although I don't see how it'd feel ugly, considering the timer API does the exact same. I suppose it's relatively easy to work around it like that, but I do feel like using string manipulation on a URL is uglier and more janky than the alternative.

The URL really should only be relevant to the target HTTP server, so that if the request really does need a specific URL hash to work, it doesn't need to be modified to identify the request.
That aside, it's also an issue for running programs concurrently, because it's possible to send a request that influences the response in another program if they send a request to the same URL.

It just seems odd to me that there's really no way to uniquely identify a request without specifically modifying the actual request's content, shouldn't every async respond have something that connects it to the place it was requested?

**Comment by SquidDev** [comment]
>  I don't see how it'd feel ugly, considering the timer API does the exact same.

So this isn't about using a unique ID — that's absolutely fine. It's about how you fit it into the current API.

The question I try to ask with this sort of thing is "if I were to redesign this from scratch, would it look like this?", and here I think it's clearly not! With the benefit of hindsight, the HTTP events should have used a unique ID *instead* of the URL.

Sometimes we have made that compromise! But I want to be really sure it's worth it, and it's not clear to me whether it is here.

**Comment by CelDaemon** [comment]
I've also felt like that about many parts of CC:T that felt inconsistent, I just felt this would be the smallest change with the largest positive impact, while also making it easy to adopt and not breaking most existing programs.

In an ideal world I'd definitely want to replace the URL instead of adding to the event some more, but right now that'd require having a separate API. I've also thought about that, but I'm not convinced having two versions of the same system side-by-side would be much better.

I hope some solution will come at some point though! I've just made the changes I would personally make on my own fork, but I haven't made a pull request in case a better solution might come to exist.

**Comment by SquidDev** [comment]
> I'm not convinced having two versions of the same system side-by-side would be much better.

Ah yeah, I definitely wouldn't propose that!

The trouble with CraftOS, is that every change requires juggling the design of the feature/API itself, backwards comparability, and whether this design allows future changes/improvements. I really struggle with this — API design is very much not my strong suit — and so I tend to be cautious with changes, only doing stuff if I'm really sure about it. I don't disagree that it's frustrating at times though!

> many parts of CC:T that felt inconsisten 

If you're willing to, could you create an issue/discussion with these? I obviously can't make any promises about fixes, but it's always useful to hear issues people are hitting.
